src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala

```diff
@@ -160,25 +162,47 @@ trait UmiConsensusCaller[C <: SimpleRead] {
   protected[umi] def sourceMoleculeId(rec: SAMRecord): String
 
   /**
-    * Converts from a SAMRecord into a SourceRead.
+    * Converts from a SAMRecord into a SourceRead.  During conversion the record is end-trimmed
+    * to remove Ns and bases below the `minBaseQuality`.  Remaining bases that are below
+    * `minBaseQuality` are then masked to Ns.
+    *
+    * @return Some(SourceRead) if there are any called bases with quality > minBaseQuality, else None
     */
-  protected def toSourceRead(rec: SAMRecord, minBaseQuality: PhredScore): SourceRead = {
+  protected[umi] def toSourceRead(rec: SAMRecord, minBaseQuality: PhredScore): Option[SourceRead] = {
+    // Extract and possibly RC the source bases and quals from the SAMRecord
     val bases = rec.getReadBases.clone()
     val quals = rec.getBaseQualities.clone()
-
     if (rec.getReadNegativeStrandFlag) {
       SequenceUtil.reverseComplement(bases)
       SequenceUtil.reverse(quals, 0, quals.length)
     }
 
+    // Mask low quality bases to Ns
     forloop (from=0, until=bases.length) { i =>
       if (quals(i) < minBaseQuality) {
         bases(i) = NoCall
         quals(i) = NoCallQual
       }
     }
 
-    SourceRead(sourceMoleculeId(rec), bases, quals)
+    // Find the last non-N base of sufficient quality in the record, starting from either the
+    // end of the read, or the end of the insert, whichever is shorter!
+    val len = {
+      var index = if (Bams.isFrPair(rec)) min(abs(rec.getInferredInsertSize), bases.length) - 1 else bases.length - 1
+      while (index >= 0 && (bases(index) == NoCall || quals(index) < minBaseQuality)) index -= 1
```

Comment:
Thinking about this more, why move the filtering and trimming to the filtering tool?  Or do you want to filter the single strand consensus reads _prior_ to duplex calling?